File: PL/3/PythonFunc/FunctionalProject.py
import sys
import re
import collections
from functools import reduce


def parse(command):
    if len(command) > 1:
        if command[1] == 'wc':
            file = command[2]
            with open(file, 'r') as myfile:
                text = myfile.read()
                count = wc(text)
                for word, freq in count.items():
                    print(word + '\t' + str(freq))
        elif command[1] == 'grep':
            file = command[3]
            with open(file, 'r') as myfile:
                text = myfile.read()
                if len(command) == 4:
                    print(grep(str(command[2]), text))
                else:
                    count = pipe(str(command[2]), text)
                    for word, freq in count.items():
                        print(word + '\t' + str(freq))
        else:
            print('Invalid Command.')
    else:
        print('Requires command line args.')


def grep(search_string, text):
    return reduce((lambda x, y: x + '\n' + y),
                  filter(lambda x: x.find(search_string) != -1,
                         text.splitlines()))


def wc(text):
    return to_dict(list(filter(None,
                               map(lambda x: re.sub("[^A-Za-z0-9]", "", x),
                                   text.lower().split()))))


def pipe(search_string, text):
    return to_dict(list(filter(None,
                               map(lambda x: re.sub("[^A-Za-z0-9]", "", x),
                                   reduce((lambda x, y: x + ' ' + y),
                                          filter(lambda x: x.find(search_string) != -1,
                                                 text.splitlines())).lower().split()))))


def to_dict(words):
    return dict(collections.Counter(words))


parse(sys.argv)

File: PL/3/JavascriptFunc/FunctionalProject.js
'use strict';
let fs = require("fs");

let grep = function(searchString, text) {
    return text.split(/\r?\n/)
        .filter(line => line.includes(searchString))
        .reduce((a, b) => (a + '\n' + b));
};

let wc = function(text) {
    return text.split(/\r?\n/)
        .map(x => x.toLowerCase())
        .reduce((a, b) => (a + ' ' + b))
        .split(/(\s+)/)
        .filter(word => word.trim().length > 0)
        .map(word => word.replace(/[^0-9a-z]/gi, ''))
        .filter(Boolean)
        .reduce(function(a, b) {
            a[b] = a.hasOwnProperty(b) ? a[b] + 1 : 1;
            return a;
            }, {});
};

let pipe = function(searchString, text) {
    return text.split(/\r?\n/)
        .filter(line => line.includes(searchString))
        .map(x => x.toLowerCase())
        .reduce((a, b) => (a + ' ' + b))
        .split(/(\s+)/)
        .filter(word => word.trim().length > 0)
        .map(word => word.replace("[^A-Za-z0-9]", ""))
        .filter(Boolean)
        .reduce(function(a, b) {
            a[b] = a.hasOwnProperty(b) ? a[b] + 1 : 1;
            return a;
        }, {});
};


let parse = function(command) {
    if (command[2] != null) {
        if (command[2] === "wc") {
            let result = wc(fs.readFileSync(command[3]).toString('utf-8'));
            for (let key in result) {
                if (result.hasOwnProperty(key)) {
                    console.log(key + "\t" + result[key]);
                }
            }
        } else if (command[2] === "grep") {
            if (command.length === 5) {
                console.log(grep(command[3],
                    fs.readFileSync(command[4]).toString('utf-8')));
            } else {
                let result = pipe(command[3],
                    fs.readFileSync(command[4]).toString('utf-8'));
                for (let key in result) {
                    if (result.hasOwnProperty(key)) {
                        console.log(key + "\t" + result[key]);
                    }
                }
            }
        } else {
            console.log("Invalid Command.");
        }
    } else {
        console.log("Requires command line args.");
    }
};

parse(process.argv);
